server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -363,11 +374,14 @@ public IndexRequest source(Map<String, ?> source) throws ElasticsearchGeneration
      *
      * @param source The map to index
      */
+    @SuppressWarnings("unchecked")
     public IndexRequest source(Map<String, ?> source, XContentType contentType) throws ElasticsearchGenerationException {
         try {
             XContentBuilder builder = XContentFactory.contentBuilder(contentType);
             builder.map(source);
-            return source(builder);
+            IndexRequest indexRequest = source(builder);
+            this.sourceAsMap = (Map<String, Object>) source;
+            return indexRequest;
```

Comment:
Maybe make a to `IngestService` instead of modifying this method (which does get invoked in other places too)?
I think we can add the following to `IngestService#innerExecute(...)` method at line 873:

```
Object rawTimestamp = ingestDocument.getSourceAndMetadata().get("@timestamp");
                if (rawTimestamp != null) {
                    indexRequest.setRawTimestamp(rawTimestamp);
                }
```

and then add a `rawTimestamp `field to `IndexRequest`.

This feels more targeted and cleaner to me.

Reply:
ok, I will modify the code as you suggest.